### PR TITLE
feat(client): error handling and test resilience

### PR DESCRIPTION
**Pull request by YoDaMa**

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -939,6 +962,7 @@ MqttClient.prototype._clearReconnect = function () {
  * @api private
  */
 MqttClient.prototype._cleanUp = function (forced, done) {
+  var that = this
```
Comment (anthonyvercolano) [inline lib/client.js:965]: >var that = this [](start = 2, length = 15)

Why did you need 'that'?

**Comment by anthonyvercolano** [comment]
>  * @param {String} type - packet type (see `protocol`)

Is this still here?

---
Refers to: lib/client.js:1012 in 8de6613. [](commit_id = 8de6613cf2a73b9f930b7c34d13857b4f7348195, deletion_comment = False)

**Comment by anthonyvercolano** [comment]
>  * @param {String} type - packet type (see `protocol`)

Still here?

---
Refers to: lib/client.js:1063 in 8de6613. [](commit_id = 8de6613cf2a73b9f930b7c34d13857b4f7348195, deletion_comment = False)

**Comment by anthonyvercolano** [comment]
>   if (packet.properties) {

Could probably use more debugs in here.  Property changes keepalive changes rc values etc...

---
Refers to: lib/client.js:1157 in 8de6613. [](commit_id = 8de6613cf2a73b9f930b7c34d13857b4f7348195, deletion_comment = False)

**Comment by anthonyvercolano** [comment]
>       // log or throw an error about unknown qos

Maybe add a debug here?

---
Refers to: lib/client.js:1268 in 8de6613. [](commit_id = 8de6613cf2a73b9f930b7c34d13857b4f7348195, deletion_comment = False)

File: lib/connect/index.js [hunk 2]
```diff
@@ -147,8 +147,9 @@ function connect (brokerUrl, opts) {
 
     return protocols[opts.protocol](client, opts)
   }
-
-  return new MqttClient(wrapper, opts)
+  var client = new MqttClient(wrapper, opts)
+  client.on('error', function () { /* Automatically set up client error handling */ })
```
Comment (anthonyvercolano) [inline lib/connect/index.js:151]: >client.on('error', function () { /* Automatically set up client error handling */ }) [](start = 2, length = 84)

Does this prevent the application from faulting if an error is emitted from mqttjs and nobody 'handles' it?

File: test/abstract_client.js [hunk 3]
```diff
@@ -170,6 +172,7 @@ module.exports = function (server, config) {
     })
 
     it('should emit end even on a failed connection', function (done) {
+      debug('client connecting.')
```
Comment (anthonyvercolano) [inline test/abstract_client.js:175]: >debug('client connecting.') [](start = 6, length = 27)

Did you want to leave these debug's in?

File: test/abstract_client.js [hunk 4]
```diff
@@ -181,21 +184,35 @@ module.exports = function (server, config) {
...
-          done(new Error('end emitted twice'))
-        })
-      })
+      // after 200ms manually invoke client.end
+      setTimeout(() => {
+        debug('manually invoking client.end')
+        var boundEnd = client.end.bind(client)
+        boundEnd()
+      }, 200)
+      // setTimeout(client.end.bind(client), 200)
```
Comment (anthonyvercolano) [inline test/abstract_client.js:193]: >// setTimeout(client.end.bind(client), 200) [](start = 6, length = 43)

Remove

File: test/abstract_client.js [hunk 5]
```diff
@@ -181,21 +184,35 @@ module.exports = function (server, config) {
...
+      }
 
-      setTimeout(client.end.bind(client), 300)
+      var spy = sinon.spy(endCallback)
+      client.on('end', spy)
+      setTimeout(() => {
+        debug('timeout invoked. Manually calling client.end()')
+        client.end.bind(client)
+        client.end()
+        // client.end.bind(client)
```
Comment (anthonyvercolano) [inline test/abstract_client.js:214]: >// client.end.bind(client) [](start = 8, length = 26)

Remove?

File: test/websocket_client.js [hunk 6]
```diff
@@ -81,7 +79,18 @@ server.on('client', function (client) {
...
+}
+
+attachWebsocketServer(server)
+
+// var serverBuilder = function () {
+//   var server = http.createServer()
+//   attachWebsocketServer(server)
+//   server.on('client', attachClientEventHandlers)
+//   return server
+// }
```
Comment (anthonyvercolano) [inline test/websocket_client.js:91]: No commented code please.

File: test/client_mqtt5.js [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+      assert.strictEqual(error.message, 'Packet has no Authentication Method')
+      // client will not be connected, so we will call done.
+      assert.isTrue(client.disconnected, 'validate client is disconnected')
+      client.end(true)
+      done()
+    })
+  })
+
+  it('auth packet', function (done) {
+    this.timeout(15000)
```
Comment (anthonyvercolano) [inline test/client_mqtt5.js:60]: >this.timeout(15000) [](start = 4, length = 19)

These are all local server connections?  Does it really need this long of a timeout?

Comment (YoDaMa) [reply test/client_mqtt5.js:60]: they are local server connections but still need this timeout.

---
In reply to: [407791930](https://github.com/mqttjs/MQTT.js/pull/1076#discussion_r407791930) [](ancestors = 407791930)

File: lib/connect/index.js (hunk 2 above)
Comment (YoDaMa) [reply lib/connect/index.js:151]: yes.

---
In reply to: [407712729](https://github.com/mqttjs/MQTT.js/pull/1076#discussion_r407712729) [](ancestors = 407712729)
